calyx-opt/src/passes/compile_sync_without_sync_reg.rs

```diff
@@ -0,0 +1,211 @@
+/// and a std_reg(1) for each barrier (`s`)
+/// It then continuously assigns the value of (`s.in`) to 1'd1 guarded by the
+/// expression that all values of `bar` for threads under the barrier are
+/// set to 1'd1
+/// Then it replaces the @sync control operator with
+/// seq {
+///     barrier;
+///     clear;
+/// }
+/// `barrier` simply sets the value of `bar` to 1'd1 and then waits
+/// for `s.out` to be up
+/// `clear` resets the value of `bar` to 1'd0 for reuse of barrier
+/// Using this method, each thread only incurs 3 cycles of latency overhead for
+/// the barrier, and we theoretically won't have a limit for number of threads
+/// under one barrier
+pub struct CompileSyncWithoutSyncReg;
+
+impl Named for CompileSyncWithoutSyncReg {
+    fn name() -> &'static str {
+        "compile-sync-without-sync-reg"
+    }
+
+    fn description() -> &'static str {
+        "Implement barriers for statements marked with @sync attribute without
+      std_sync_ref"
+    }
+}
+
+// Data structure storing the shared `s` register and the guard accumulator
+// for guarding `s.in`
+type BarrierMap = HashMap<u64, (RRC<ir::Cell>, Box<ir::Guard<ir::Nothing>>)>;
+
+trait Gettable {
+    fn get_reg(&mut self, idx: &u64) -> &mut RRC<ir::Cell>;
+
+    fn get_guard(&mut self, idx: &u64) -> &mut Box<ir::Guard<ir::Nothing>>;
+}
+
+impl Gettable for BarrierMap {
+    fn get_reg(&mut self, idx: &u64) -> &mut RRC<ir::Cell> {
+        let (cell, _) = self.get_mut(idx).unwrap();
+        cell
+    }
+
+    fn get_guard(&mut self, idx: &u64) -> &mut Box<ir::Guard<ir::Nothing>> {
+        let (_, gd) = self.get_mut(idx).unwrap();
+        gd
+    }
+}
+
+// instantiates the hardware and the two groups: `bar` and `clear` for each
+// barrier
+fn build_barrier_group(
+    builder: &mut ir::Builder,
+    barrier_idx: &u64,
+    barrier_reg: &mut BarrierMap,
+) -> ir::Control {
+    let group = builder.add_group("barrier");
+    structure!(
+        builder;
+        let bar = prim std_reg(1);
+        let z = constant(0, 1);
+        let constant = constant(1, 1);
+    );
+
+    barrier_reg
+        .get_guard(barrier_idx)
+        .update(|g| g.and(guard!(bar["out"])));
+
+    let s = barrier_reg.get_reg(barrier_idx);
+
+    let assigns = build_assignments!(builder;
+        bar["in"] = ? constant["out"];
+        bar["write_en"] = ? constant["out"];
+        group["done"] = ? s["out"];
+    );
+    group.borrow_mut().assignments.extend(assigns);
+
+    let clear = builder.add_group("clear");
+    let clear_assigns = build_assignments!(builder;
+        bar["in"] = ? z["out"];
+        bar["write_en"] = ? constant["out"];
+        clear["done"] = ? bar["done"];);
+    clear.borrow_mut().assignments.extend(clear_assigns);
+
+    let stmts = vec![ir::Control::enable(group), ir::Control::enable(clear)];
+
+    ir::Control::seq(stmts)
+}
+
+// produces error if `invoke` or `enable` are marked with @sync
+fn produce_err(con: &ir::Control) -> CalyxResult<()> {
```

Comment:
A simpler way to do this is:
```
if let ir::Control::Empty(_) { 
  Ok(())
} else {
  if con.has_attribute("sync") {
    return Err(...)
  }
}
```

Reply:
Turns out this might not be the simpler way? Because we would want to access the `attribute` field in each variant of the `Control` enum.